### PR TITLE
Moving away from deprecated i/u suffixes in libcore

### PR DESCRIPTION
**Pull request by alfiedotwtf**

### CODE REVIEW

**Comment by alfie** [comment]
If everyone is happy with this change, I'll move forward with the rest. Otherwise, let me know what needs to be done and I'll give it a go.

File: src/libcore/char.rs [hunk 1]
```diff
@@ -102,7 +102,7 @@ pub fn from_digit(num: uint, radix: uint) -> Option<char> {
             if num < 10 {
                 Some(transmute(('0' as uint + num) as u32))
             } else {
-                Some(transmute(('a' as uint + num - 10u) as u32))
```
Comment (eddyb) [inline src/libcore/char.rs]: Unnecessary suffixes.

**Comment by eddyb** [comment]
I started looking through these and realized commenting on each one would be pointless.
Most uses of integral suffixes outside of code examples in documentation seem to be unnecessary.
Do we have any style guidelines related to the use of suffixes in general? cc @aturon

**Comment by alfie** [comment]
@eddyb I've tried looking for style on this but didn't find anything. The word on IRC today was to drop the suffix when it can be inferred, which I've done.

Are you sure you want to drop most suffixes? e.g.

    let mut n = 0i;   // original: int
    let mut n = 0is; // mine: isize
    let mut n = 0;   //  dropped: i32

I deliberately didn't want to change the datatypes like above, only move away from the deprecated suffixes.

**Comment by eddyb** [comment]
@alfie the type only changes if it's unconstrained anywhere, which seems a bit suspicious. Do you have examples of this? I'll look through the diff again, I guess.

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -846,7 +846,7 @@ macro_rules! tuple {
             fn fmt(&self, f: &mut Formatter) -> Result {
                 try!(write!(f, "("));
                 let ($(ref $name,)*) = *self;
-                let mut n = 0i;
+                let mut n = 0is;
```
Comment (eddyb) [inline src/libcore/fmt/mod.rs]: First usage that needs annotation. Even then, this is used for 3 states, which would be only 2 if we were to special-case one-element tuples. Feels saner that way IMO.
We can do it without adding another case to the macro, too, *and* while removing the `peel` helper macro.
Something like `$first, $($name),*`.

File: src/libcore/num/mod.rs [hunk 3]
```diff
@@ -1571,8 +1571,8 @@ macro_rules! from_str_radix_float_impl {
                 let exp = match exp_info {
                     Some((c, offset)) => {
                         let base = match c {
-                            'E' | 'e' if radix == 10 => 10u as $T,
-                            'P' | 'p' if radix == 16 => 2u as $T,
+                            'E' | 'e' if radix == 10 => 10us as $T,
+                            'P' | 'p' if radix == 16 => 2us as $T,
```
Comment (eddyb) [inline src/libcore/num/mod.rs]: Use `10.0` and `2.0` and you can remove both the suffix and the cast.

File: src/libcore/slice.rs [hunk 4]
```diff
@@ -653,7 +653,7 @@ macro_rules! iterator {
                             self.ptr = transmute(self.ptr as uint + 1);
 
                             // Use a non-null pointer value
-                            Some(transmute(1u))
+                            Some(transmute(1us))
```
Comment (eddyb) [inline src/libcore/slice.rs]: This should be a dummy constant somewhere - which I thought we had.

**Comment by eddyb** [comment]
None of the 3 cases I found would actually cause a subtle change in semantics (the last one would error if `1` would default to a differently-sized type).
They are *all* in macros, and even then, the suffixes can be removed while cleaning up the way the macro works.
I ignored examples in doc comments because they are highly artificial, though some of them may work with less suffixes.

**Comment by alfie** [comment]
@eddyb Ok, no problem. I've updated the patch:

    - Suffixes have been dropped in code
    - Suffixes updated from deprecated i/u in documentation

I'd like for the examples to be updated because it's going to be confusing in the future once i/u are no longer valid syntax.

File: src/libcore/num/mod.rs [hunk 5]
```diff
@@ -1571,8 +1571,8 @@ macro_rules! from_str_radix_float_impl {
                 let exp = match exp_info {
                     Some((c, offset)) => {
                         let base = match c {
-                            'E' | 'e' if radix == 10 => 10u as $T,
-                            'P' | 'p' if radix == 16 => 2u as $T,
+                            'E' | 'e' if radix == 10 => 10.0 as $T,
+                            'P' | 'p' if radix == 16 => 2.0 as $T,
```
Comment (eddyb) [inline src/libcore/num/mod.rs]: You don't need `as $T` at all, the type is either `f32` or `f64`.

**Comment by alfie** [comment]
@eddyb Thanks for that. I wasn't too sure about that one.

**Comment by huonw** [comment]
I think much of the documentatio can have the suffixes dropped too: in most cases it doesn't matter if it's a `i32` or `isize` or whatever, and the cases where it does matter it should infer (I would guess).

**Comment by alfie** [comment]
@huonw No problem. Fixed.

File: src/libcore/default.rs [hunk 6]
```diff
@@ -150,13 +150,13 @@ default_impl! { (), () }
 default_impl! { bool, false }
 default_impl! { char, '\x00' }
 
-default_impl! { uint, 0u }
+default_impl! { uint, 0us }
```
Comment (eddyb) [inline src/libcore/default.rs]: All of these suffixed integers should be able to be just `0`.

**Comment by alfie** [comment]
@eddyb Thanks for the clarification. I wasn't too sure about that one. Fixed.

**Comment by eddyb** [comment]
I just checked and the only `us`/`is` suffixes added are in `default_impl!` invocations and the `transmute` calls in the slice iterators - both of those can be removed.
I am honestly surprised this possible on such a scale. Great work, @alfie!

**Comment by alfie** [comment]
@eddyb Thanks for that. I wasn't sure what to replace it with, but that looks reasonable. Fixed.

**Comment by huonw** [comment]
@bors: r+ 4

Looks great!

**Comment by huonw** [comment]
@bors: r+ 45b6d40

(Drat, a single letter is too short. :stuck_out_tongue_closed_eyes: )

**Comment by alfie** [comment]
Was that auto-mac-32-opt failure legit?

File: src/libcore/iter.rs [hunk 7]
```diff
@@ -1138,7 +1138,7 @@ pub trait AdditiveIterator<A> {
     /// ```
     /// use std::iter::AdditiveIterator;
     ///
-    /// let a = [1i, 2, 3, 4, 5];
+    /// let a = [1, 2, 3, 4, 5];
```
Comment (huonw) [inline src/libcore/iter.rs]: I think this needs a suffix (e.g. `1i32`).

File: src/libcore/option.rs [hunk 8]
```diff
@@ -384,9 +384,9 @@ impl<T> Option<T> {
     /// # Example
     ///
     /// ```
-    /// let k = 10u;
-    /// assert_eq!(Some(4u).unwrap_or_else(|| 2 * k), 4u);
-    /// assert_eq!(None.unwrap_or_else(|| 2 * k), 20u);
+    /// let k = 10;
```
Comment (huonw) [inline src/libcore/option.rs]: And so does this.

**Comment by huonw** [comment]
@alfie yeah, I think so.

**Comment by alfie** [comment]
@huonw Thanks. Ok, will fix.

Can you just clear something up for me please. These are all in comments and not code, but it still made the build fail? 

Also, you said to make the following to an i32:

   -    /// let a = [1i, 2, 3, 4, 5];
   +    /// let a = [1, 2, 3, 4, 5];

I thought that if an integer doesn't have a suffix, it's typed as an i32 as a default? If so, why should this need an i32 suffix?

**Comment by huonw** [comment]
The `rustdoc` supports [running code examples in doc-comments as tests](http://doc.rust-lang.org/nightly/book/documentation.html#testing-the-documentation) and we run these tests on all crates to keep the docs up to date.

I'm not sure what's happening with those examples, but inference can be a little sensitive/less-than-perfect when there's closures around, at the moment.

**Comment by alfie** [comment]
@huonw Thanks for the explanation.

File: src/libcore/char.rs [hunk 9]
```diff
@@ -258,7 +258,29 @@ impl CharExt for char {
     #[inline]
     #[unstable = "pending decision about Iterator/Writer/Reader"]
     fn encode_utf8(self, dst: &mut [u8]) -> Option<uint> {
-        encode_utf8_raw(self as u32, dst)
+        // Marked #[inline] to allow llvm optimizing it away
```
Comment (huonw) [inline src/libcore/char.rs]: Hm, why was this changed?

**Comment by alfie** [comment]
I got a conflict while rebasing onto master. I was sure I resolved it correctly. Fixing now :(

**Comment by alfie** [comment]
Apologies. I resolved the conflict that the wrong way. Fixed.

**Comment by huonw** [comment]
@bors: r+ f67e747

It's not problem at all, rebases are tricky like that sometimes.
